File: alchemist/alchemist-incarnation-scafi/src/test/scala/it/unibo/alchemist/scafi/test/TestInSimulator.scala
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */

package it.unibo.alchemist.scafi.test

/*******************************************************************************
 * Copyright (C) 2010-2018, Danilo Pianini and contributors listed in the main
 * project's alchemist/build.gradle file.
 * 
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception, as described in the file
 * LICENSE in the Alchemist distribution's top directory.
 ******************************************************************************/
import java.io.InputStream

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import it.unibo.alchemist.core.implementations.Engine
import it.unibo.alchemist.loader.YamlLoader
import it.unibo.alchemist.model.implementations.molecules.SimpleMolecule
import it.unibo.alchemist.model.interfaces.{Environment, Position}
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FunSuite, Matchers}

import scala.collection.JavaConverters.{asScalaBufferConverter, mapAsScalaMapConverter}

@SuppressFBWarnings(value = Array("SE_BAD_FIELD"), justification="We are not going to Serialize test classes")
@RunWith(classOf[JUnitRunner])
class TestInSimulator[P <: Position[P]] extends FunSuite with Matchers {
  
  test("Basic test"){
    testNoVar("/plain_vanilla.yml")
  }
  test("Gradient"){
    val env = testNoVar[Any]("/test_gradient.yml")
    env.getNodes.asScala.foreach(node => {
      val contents = node.getContents.asScala
      (contents.get(new SimpleMolecule("it.unibo.alchemist.scafi.test.ScafiGradientProgram")).get.asInstanceOf[Double]) should (be >= 0.0 and be <= 100.0)
    })
  }

  test("Environment"){
    val env = testNoVar[Any]("/test_env.yml")
    env.getNodes.asScala.foreach(node => {
      val contents = node.getContents.asScala
      def inputMolecule = (contents.get(new SimpleMolecule("number")).get.asInstanceOf[Int])
      def outputMolecule = (contents.get(new SimpleMolecule("number2")).get.asInstanceOf[Int])
      if(node.getId==0) {
        inputMolecule shouldBe (77)
        outputMolecule shouldBe (177)
      } else {
        inputMolecule shouldBe (-500)
        outputMolecule shouldBe (-400)
      }
    })
  }

  private def testNoVar[T](resource: String, maxSteps: Long = 1000): Environment[T, P] = {
    testLoading(resource, Map(), maxSteps)
  }

  private def testLoading[T](resource: String, vars: Map[String, java.lang.Double], maxSteps: Long = 1000): Environment[T, P] = {
    import scala.collection.JavaConverters._
    val res: InputStream = classOf[TestInSimulator[P]].getResourceAsStream(resource)
    res shouldNot be(null)
    val env: Environment[T, P] = new YamlLoader(res).getWith(vars.asJava)
    val sim = new Engine[T, P](env, maxSteps)
    sim.play()
    sim.run()
    if(sim.getError.isPresent) throw new Exception(sim.getError.get().getMessage)
    env
  }
}

File: alchemist-incarnation-protelis/src/test/java/it/unibo/alchemist/model/protelis/TestInSimulator.java
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.protelis;

import com.google.common.collect.Maps;
import it.unibo.alchemist.core.Engine;
import it.unibo.alchemist.core.Simulation;
import it.unibo.alchemist.boundary.LoadAlchemist;
import it.unibo.alchemist.model.Environment;
import it.unibo.alchemist.model.Position;
import org.apache.commons.io.FileUtils;
import org.jooq.lambda.Unchecked;
import org.junit.jupiter.api.Test;
import org.kaikikm.threadresloader.ResourceLoader;

import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.file.Files;
import java.util.Collections;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.FutureTask;

import static org.junit.jupiter.api.Assertions.assertNotNull;

/**
 * A series of tests checking that our Yaml Loader is working as expected.
 */
class TestInSimulator {

    /**
     * Basic loading capabilities.
     */
    @Test
    void testBase() {
        testNoVar("testbase.yml");
    }

    /**
     * Tests loading custom nodes.
     */
    @Test
    void testCustomNodes() {
        testNoVar("customnodes.yml");
    }

    /**
     * Test the ability to load a Protelis module from classpath.
     */
    @Test
    void testLoadProtelisModule() {
        testNoVar("test00.yml");
    }

    /**
     * Test the ability to load Protelis modules that are dynamically added and removed from classpath in a multithreaded system.
     *
     * @throws IOException causes failure
     * @throws URISyntaxException causes failure
     * @throws ExecutionException causes failure
     * @throws InterruptedException causes failure
     */
    @Test
    void testThreadDependentLoadModule() throws IOException, URISyntaxException, ExecutionException, InterruptedException {
        final Callable<Void> c = () -> {
            testNoVar("18-export.yml");
            return null;
        };
        final File d1 = createDependenciesDirectory();
        try {
            final FutureTask<Void> ft1 = new FutureTask<>(c);
            final Thread t1 = new Thread(ft1);
            t1.setContextClassLoader(new URLClassLoader(new URL[] {d1.toURI().toURL()}));
            t1.start();
            ft1.get();
        } finally {
            FileUtils.deleteDirectory(d1);
        }
        final File d2 = createDependenciesDirectory();
        try {
            final FutureTask<Void> ft2 = new FutureTask<>(c);
            final Thread t2 = new Thread(ft2);
            t2.setContextClassLoader(new URLClassLoader(new URL[] {d2.toURI().toURL()}));
            t2.start();
            ft2.get();
        } finally {
            FileUtils.deleteDirectory(d2);
        }
    }

    private File createDependenciesDirectory() throws IOException, URISyntaxException {
        final File d = Files.createTempDirectory("alchemist-test").toFile();
        FileUtils.copyDirectory(new File(ResourceLoader.getResource("advancedorig").toURI()), new File(d, "advanced"));
        FileUtils.copyDirectory(new File(ResourceLoader.getResource("plutoorig").toURI()), new File(d, "pluto"));
        return d;
    }

    /**
     * Test the ability to inject variables.
     */
    @Test
    void testLoadWithVariable() {
        final Map<String, Double> map = Maps.newLinkedHashMap();
        map.put("testVar", 10d);
        testLoading("test00.yml", map);
    }

    private static void testNoVar(final String resource) {
        testLoading(resource, Collections.emptyMap());
    }

    private static <T, P extends Position<P>> void testLoading(final String resource, final Map<String, Double> vars) {
        final var res = ResourceLoader.getResource(resource);
        assertNotNull(res, "Missing test resource " + resource);
        final Environment<T, P> environment = LoadAlchemist.from(res).<T, P>getWith(vars).getEnvironment();
        final Simulation<T, P> sim = new Engine<>(environment, 10_000);
        sim.play();
//        if (!java.awt.GraphicsEnvironment.isHeadless()) {
//            it.unibo.alchemist.boundary.gui.SingleRunGUI.make(sim);
//        }
        sim.run();
        sim.getError().ifPresent(Unchecked.consumer(e ->  {
            throw e;
        }));
    }

}
